### ISSUE TITLE
How to convert emu saves to eeprom bin?

### ISSUE THREAD

**Issue by LDAsh72** [fgsfdsfgs/perfect_dark#233]
I posted this:-
https://github.com/fgsfdsfgs/perfect_dark/issues/232
...in the hope that an answer to that would be given.
The other save file that was posted instead seems to strongly indicate that it's possible, but I haven't seen it mentioned anywhere?

**Comment by fgsfdsfgs** [comment]
Oh. No, there's no tool to do that, Graslu just got that save on the port.
It is possible to write a tool that does this though.

**Comment by LDAsh72** [comment]
Wow, he did all of it in just 4 hours and somehow got 100s of headshots in combat simulator with auto-aim turned on.  I assumed it was GameShark tricks and save-scumming, but obviously he's a pro speedrunning genius.  That's cool.

**Comment by fgsfdsfgs** [comment]
Are you implying I'm withholding a save converter from you or something? I really don't know and frankly don't care how exactly he got it, I just posted it because it's a 100% save that you can use if you want until there's a save converter available.

**Comment by megatog615** [comment]
> Wow, he did all of it in just 4 hours and somehow got 100s of headshots in combat simulator with auto-aim turned on. I assumed it was GameShark tricks and save-scumming, but obviously he's a pro speedrunning genius. That's cool.

There's a strong indication that Graslu's save file was generated by the port but hex-edited to unlock as much as possible. Lots of the records indicate times in 0 seconds.

Chill out.

**Comment by fgsfdsfgs** [comment]
Fair enough. I just assumed he unlocked everything while streaming it or something, not that it really matters.
Saves can be converted from emulator to port by byteswapping all the fields, so a tool would be relatively simple to write.

**Comment by Graslu** [comment]
All the cheats, challenges, gun range and co-op were completed legit. ~~Maps show 0 seconds in times because that's my save from when the port would crash by loading Villa, so I was given a build with all the maps unlocked by default.~~ <- This was the earliest 100% save, I just completed it on my Steam Deck on and off. The final 100% save file I shared has everything unlocked manually.

No tool was used to convert the save from emulator to the PC port.

**Comment by Bobbar** [comment]
FWIW, I just modified some of the game code to force cheats/levels/guns/etc to be unlocked.  I took some of the "hacks" that are applied by the debug mode and just put them in permanently.    It shouldn't be too much trouble to just add an option to the port to unlock everything.

I'll take a look and see if I think I can implement a cleaner pattern which would allow all unlockables to be toggled from the port menu and make a PR when I have the time.

**Comment by LuisMayo** [comment]
> Saves can be converted from emulator to port by byteswapping all the fields, so a tool would be relatively simple to write.

I guess you're talking about endianness

Any documentation on the byte offsets for the fields?

BTW thanks for the port. I was already playing this in 1964GE but native support is always better ♥️

**Comment by fgsfdsfgs** [comment]
Check `struct pakfileheader` in `types.h`. I think the rest of the save file is a bitstream, so it might only be the header that needs swapping. Note that one EEPROM file can contain multiple saves.

**Comment by Atari-Dude** [comment]
Raphnet N64 controller adapters can read-out controller paks but could it be possible allow the port to read .MPKs? Even if you can't save back onto them for some reason, just being able to continue to use my preserved profiles from 16 years of accumulated playing would be a dream come true. Or even just being able to manually adjust my player's stats to what I had before would suffice too lol 😅

**Comment by gittybittycomboe** [comment]
Where are save files located? Within the ROM itself? I want to back them up.

**Comment by fgsfdsfgs** [comment]
On Windows usually in the same folder as `pd.exe`. On Linux usually in `~/.local/share/perfectdark`. They're in a file called `eeprom.bin`.

**Comment by PhobosAnomaly666** [comment]
This is a feature I'd really like to see implemented (or have some sort of doc on how to do it - I used Hex editors in the past but I'm just good for reading instructions, can't understand the logic - yet).

Loving the work done on every aspect, this would be the icing on the cake.

**Comment by MaikelChan** [comment]
In case someone is interested, I made a save editor which is also capable of converting between N64 eeprom to PC format and viceversa. It also takes into account that the PC port added the classic weapons in multiplayer, which made weapons setup incompatible between both versions.

Just open the save in the desired format, go to the menu `Tools`, change the `Save Format`, and `Save As...` with a new name.

https://github.com/MaikelChan/PDSaveEditor

**Comment by PhobosAnomaly666** [comment]
What a legend.

Giving this a try ASAP, thanks a lot !

**Comment by LonelySpaceDetective** [comment]
> In case someone is interested, I made a save editor which is also capable of converting between N64 eeprom to PC format and viceversa. It also takes into account that the PC port added the classic weapons in multiplayer, which made weapons setup incompatible between both versions.
> 
> Just open the save in the desired format, go to the menu `Tools`, change the `Save Format`, and `Save As...` with a new name.
> 
> https://github.com/MaikelChan/PDSaveEditor

Oh this looks incredibly handy for Combat Simulator setups, appreciate it.

**Comment by PhobosAnomaly666** [comment]
> In case someone is interested, I made a save editor which is also capable of converting between N64 eeprom to PC format and viceversa. It also takes into account that the PC port added the classic weapons in multiplayer, which made weapons setup incompatible between both versions.
> 
> Just open the save in the desired format, go to the menu `Tools`, change the `Save Format`, and `Save As...` with a new name.
> 
> https://github.com/MaikelChan/PDSaveEditor

I can confirm this worked like a charm for my 17 years old savegame for the European rom, played through various emulators since my teenage years... This is amazing.

Wonderful job on this tool, it definitely should be advertised on the main page for the decomp because this is something many people would find useful considering the multiplayer side of the game stores a lot of cool informations.

**Comment by creamcast** [comment]
> In case someone is interested, I made a save editor which is also capable of converting between N64 eeprom to PC format and viceversa. It also takes into account that the PC port added the classic weapons in multiplayer, which made weapons setup incompatible between both versions.
> 
> Just open the save in the desired format, go to the menu `Tools`, change the `Save Format`, and `Save As...` with a new name.
> 
> https://github.com/MaikelChan/PDSaveEditor

I needed to edit my PC save and used your tool but if i open an eeprom made with the PC version and try to modify it, it doesn't actually alter the contents of the file. I could get it to work by first making a save on an emulator for the n64 version then opening -> editng it, saving it as PC file, then it worked. Something in the PC version must be making the save files incompatible, just thought I'd let you know.

**Comment by MaikelChan** [comment]
> I needed to edit my PC save and used your tool but if i open an eeprom made with the PC version and try to modify it, it doesn't actually alter the contents of the file. I could get it to work by first making a save on an emulator for the n64 version then opening -> editng it, saving it as PC file, then it worked. Something in the PC version must be making the save files incompatible, just thought I'd let you know.

Can you share that PC save file that didn't work for you? Thanks!
